Set `local_size` in reduction to default value

nilsfriess opened the pull request:
Passing a `local_size` of 1 in a `parallel_for` reduction previously caused an infinite loop. This PR sets the `local_size` to a default value of 128 if the user passed 1.

Fixes #857

nilsfriess committed:
Set `local_size` in reduction to default value 128 if user passed 1

A `local_size` of 1 was previously causing an infinite loop.

illuhad commented on include/hipSYCL/glue/generic/hiplike/hiplike_kernel_launcher.hpp:
```diff
@@ -420,6 +420,11 @@ determine_reduction_stages(sycl::range<Dimensions> global_size,
+  for (int i=0; i < Dimensions; ++i)
+    // Check if local_size == 1, this would cause an infinite loop
+    if (local_size[i] == 1)
+      local_size[i] = 128;
```
So, if Dimensions is greater than 1, we end up with (128, 128, ...) as work group size. This probably will break on CUDA and HIP, since those devices typically support at most 1024 elements in a work group.

Wouldn't it be sufficient to execute the reduction kernel always as a one-dimensional thing?

nilsfriess replied:
Yes that makes sense, I didn't really think about what happens if `Dimensions > 1`. But I don't quite understand what you mean by executing the kernel as a 1D thing

illuhad replied:
I meant just executing the kernel with a 1 dimensional parallel iteration space, i.e. using as a group size something like `(size, 1, 1)` if `Dimensions == 3`.
(Please check whether it needs to be `(size, 1, 1)` or `(1, 1, size)` - SYCL works in terms of `(1, 1, size)`, but CUDA/HIP in terms of (`size, 1, 1)`, so we flip the indices somewhere in the kernel launcher).

nilsfriess replied:
I'm not sure if I understand. 

Say we have a reduction over a `nd_range<2>({N, M}, {1, 16})`.  Then in the first dimension we still have the problem that computing the reduction stage sizes ends in an infinite loop (because the number of groups in the first step will be computed as `num_groups[0] = N / local_size[0]`, then this will again be divided by `local_size[0]` etc. until `num_groups[0] == 1` but `local_size[0] == 1` so always `num_groups[0] == N`).

illuhad replied:
That's true, but I think this problem only affects when we launch dedicated reduction kernels after the user-provided kernels here:
https://github.com/illuhad/hipSYCL/blob/56df01807eeb81cde76c4e0d6266d78e365ccf24/include/hipSYCL/glue/generic/hiplike/hiplike_kernel_launcher.hpp#L852

I think these could be run as 1D, since all they have to do is collect values from previous kernels and reduce them further.
In fact, looking at the code, they seem to be invoked as 1D already - so can we maybe just somehow remove the `Dimensions > 1` code paths in `determine_reduction_stages()`?.
I think all that code actually needs to care about is the total group size (`x * y * z`) so that it knows how many items will be reduced to one.

nilsfriess replied:
Okay, I see, makes sense :)

I've changed it so that ranges of the form `(N,M,L)` are transformed to `(N*M*L,1,1)` (or `(N,M)` to `(N*M,1)` in case of 2D) before computing the reduction stages.

illuhad commented:
Thinking about it some more, would it maybe make sense to add the reduction with group size of 1 to our test cases?

nilsfriess committed:
Add test with timeout for reduction with `local_size = 1`

nilsfriess committed:
Check if boost can stop tests for test that could end in infinite loop

nilsfriess commented on tests/sycl/reduction.cpp:
```diff
@@ -327,6 +328,12 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(two_reductions, T, large_test_types) {
 }
 
+#ifdef BOOST_SIGACTION_BASED_SIGNAL_HANDLING
+BOOST_AUTO_TEST_CASE(local_size_one_reduction, * boost::unit_test::timeout(2)) {
```
I have added a test case that has a timeout of 2 seconds to detect if the reduction is stuck in an infinite loop.

The `#ifdef` checks if boost is actually able to force stop the test if it times out, otherwise the test would be reported as failed but the test suite would still continue execution (in the infinite loop). I don't know however when `BOOST_SIGACTION_BASED_SIGNAL_HANDLING` is actually defined, on my setup on jarvis, for instance, it's not. So I'm not sure if this test makes sense this way.

nilsfriess committed:
Reduce group/local range to 1D before computing reduction stages

nilsfriess committed:
Also reduce computed number of group size to 1D

nilsfriess committed:
Add tests for 2D and 3D reductions

nilsfriess commented:
I've now also added tests for two- and three-dimensional reductions (both "general" reductions as well as reductions with `local_size = 1`. I also removed the the timeout test described above again, I guess it doesn't really make sense to have such a test if it isn't even supported on the system we're mostly using to test things.

nilsfriess committed:
Transform indices to (1,1,N*M*L) instead of (N*M*L,1,1)

CUDA/HIP uses n-dimensional indices of the form (N*M*L,1,1) but the
indices are reversed later in the kernel launcher, so here we have to
order the indices as (1,1,N*M*L).

nilsfriess committed:
Use correct index when checking if `local_size == 1`

nilsfriess committed:
Use loop indices from 0..Dim-2 instead of 1..Dim-1
